design/external-authorization-design.md

```diff
@@ -0,0 +1,669 @@
+# External Authorization Support
+
+Status: Reviewing
+
+## Abstract
+This document describes a design for performing request authorization for virtual hosts hosted by Contour.
+
+## Background
+
+- [#432][5] Provide support for integrating with an external auth service.
+- [#2459][3] External auth background and goals.
+- [#2325][9] Failures in External services configured in envoy should be visible to the application developer.
+
+## Goals
+- Allow operators to integrate existing authorization services with Contour.
+- Allow arbitrary types of authentication and authorization to be supported.
+- Decouple Contour from external authorization, so it can evolve at an independent rate.
+- Integrate cleanly with the HTTPProxy API.
+
+## Non Goals
+
+- Support for multiple external authorization mechanisms (i.e. the protocol the proxy uses to communicate with the authorization server).
+  Various proxies have different mechanisms for integrating external services, but it is impractical to support them all.
+
+- Abstracting the Envoy authorization mechanism.
+  The scope of abstracting the Envoy authorization mechanism is too large to be tractable.
+  Abstracting the protocol would also work against the goal of being able to integrate existing authorization servers.
+
+## High-Level Design
+A new `SupportService` CRD adds a way to represent and track an authorization service.
+This CRD is relatively generic, so that it can be re-used for Envoy rate limiting and logging services.
+The core of the `SupportService` CRD is a `projectcontour.v1.HTTPProxy` `Service` specification.
```

Comment:
This part will need updating if we make a new struct, I would propose something like:
```suggestion
The core of the `SupportService` CRD is a subset of the `projectcontour.v1.HTTPProxy` `Service` specification.
```